### PR TITLE
`Add 'ancestorOrigins' property to the LocationWrapper` (close #1342)

### LINKED ISSUE #1342: Add 'ancestorOrigins' property to the LocationWrapper
https://html.spec.whatwg.org/multipage/history.html#dom-location-ancestororigins

Access to the property found in https://www.wildberries.ru/catalog/4287053/detail.aspx?targetUrl=GP

### PR DESCRIPTION
**Pull request by Farfurix**
https://github.com/DevExpress/testcafe-hammerhead/issues/1342

### Changes
1. Add the `ancestorOrigins` property to `LocationWrapper` and wrap its returned `DOMStringList` result.

### Notes
https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#the-domstringlist-interface:
> 2.7.3 The DOMStringList interface
> The DOMStringList interface is a non-fashionable retro way of representing a list of strings.
> ...
> New APIs must use `sequence<DOMString>` or equivalent rather than `DOMStringList`.

### CODE REVIEW

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
Don't retest. Waiting for https://github.com/DevExpress/testcafe-hammerhead/pull/1801PR will be merged.

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

**Comment by Farfurix** [comment]
@testcafe-build-bot retest

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+function updateOrigin (wrapper, idx, newOrigin) {
+    wrapper[idx] = newOrigin;
+
+    nativeMethods.objectDefineProperty.call(Object, wrapper, idx, {
+        value:        newOrigin,
+        configurable: true,
+        enumerable:   true,
+        writable:     false
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: `false` is default value

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        if (isCrossDomainParent) {
+            getCrossDomainOrigin(parentWindow, origin => {
+                updateOrigin(this, i, origin);
+            });
+        }
+    }
+}
+
+DOMStringListWrapper.prototype = nativeMethods.objectCreate.call(Object, DOMStringList.prototype);
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: Let's try to use:
```js
DOMStringListWrapper.prototype = nativeMethods.objectCreate(DOMStringList.prototype);
```

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+DOMStringListWrapper.prototype.item = function (index) {
+    return this[index];
+};
+
+DOMStringListWrapper.prototype.contains = function (string) {
+    if (typeof string !== 'string')
+        string = String(string);
+
+    for (let i = 0; i < this._nativeLength; i++) {
+        if (this.item(i) === string)
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: I think the function call is unnessesary, just use `this[i]`

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    }
+}
+
+DOMStringListWrapper.prototype = nativeMethods.objectCreate.call(Object, DOMStringList.prototype);
+
+DOMStringListWrapper.prototype.item = function (index) {
+    return this[index];
+};
+
+DOMStringListWrapper.prototype.contains = function (string) {
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: rename argument to `origin`

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+nativeMethods.objectDefineProperty.call(Object, DOMStringListWrapper.prototype, 'length', {
+    configurable: true,
+    enumerable:   true,
+    get:          function () {
+        return this._nativeLength;
+    }
+});
+
+function updateOrigin (wrapper, idx, newOrigin) {
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: `idx` -> `index`

File: src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js (hunk 2 above)
Comment (LavrovArtem) [reply src/client/sandbox/code-instrumentation/location/ancestor-origins-wrapper.js]: And there https://github.com/DevExpress/testcafe-hammerhead/pull/1804/files#diff-d04bb877945ffff88eb1829bf6627b47R44

File: src/client/sandbox/code-instrumentation/location/wrapper.js [hunk 6]
```diff
@@ -135,6 +137,40 @@ export default class LocationWrapper extends EventEmitter {
             }
         }));
 
+        if (window.location.ancestorOrigins) {
+            this.messageIdGenerator  = createIntegerIdGenerator();
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/wrapper.js]: is unused line, remove it

File: src/client/sandbox/code-instrumentation/location/wrapper.js [hunk 7]
```diff
@@ -135,6 +137,40 @@ export default class LocationWrapper extends EventEmitter {
...
+            this.GET_ORIGIN_CMD      = 'hammerhead|command|get-origin';
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/wrapper.js]: Move the `GET_ORIGIN_CMD` and `ORIGIN_RECEIVED_CMD` after the imports block as const

File: src/client/sandbox/code-instrumentation/location/wrapper.js [hunk 8]
```diff
@@ -135,6 +137,40 @@ export default class LocationWrapper extends EventEmitter {
...
+            this.ORIGIN_RECEIVED_CMD = 'hammerhead|command|origin-received';
+
+            const cbs         = Object.create(null);
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/wrapper.js]: `cbs` -> `callbacks`
and use nativeMethods

File: src/client/sandbox/code-instrumentation/location/wrapper.js [hunk 9]
```diff
@@ -135,6 +137,40 @@ export default class LocationWrapper extends EventEmitter {
...
+            const idGenerator = createIntegerIdGenerator();
+
+            const getCrossDomainOrigin = (win, cb) => {
```
Comment (LavrovArtem) [inline src/client/sandbox/code-instrumentation/location/wrapper.js]: `cb` -> `callback`

File: test/client/data/cross-domain/get-ancestor-origin.html [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    var postMessage = hammerhead.sandbox.event.message.postMessage;
+
+    var getLocation     = window[INSTRUCTION.getLocation];
+    var locationWrapper = getLocation(window.location);
+    var ancestorOrigins = locationWrapper.ancestorOrigins;
+
+    window.onmessage = function (evt) {
+        if (evt.data = 'get ancestorOrigin') {
+            var data = {
+                msg: ancestorOrigins[0]
```
Comment (LavrovArtem) [inline test/client/data/cross-domain/get-ancestor-origin.html]: Send `JSON.stringify(ancestorOrigins)` for the length checking

**Pull request merged**